### ISSUE TITLE
Submission line numbers off by one

### ISSUE THREAD

**Issue by codesman** [exercism/exercism.io#3050]
I just noticed this on my recent submission: http://exercism.io/submissions/80a830a9ac254338ae78991604c085b5

The line numbers do not line up with the code. The code starts before the line numbers start and ends after the line numbers end.

At first it looks like it might be a css issue, but it's not.  In the submission I linked to, there are 120 line numbers, but the file actually has 121 lines, so the app is not correctly parsing the number of lines in the file or the view is not presenting them all.

I edited the source in the inspector and added an element for line 121 and everything lines up great if that is there, so just need to figure out why it's off by one...

**Comment by kytrinyx** [comment]
I merged https://github.com/exercism/exercism.io/pull/3048 today, which seems to be the likely cause.

@k3rni would you mind taking a look at this?

**Comment by k3rni** [comment]
Using @codesman's code as a sample test input, I took a quick look. It's very confusing, because it does render properly in tests:

```html
<span id="L119">        <span class="k">return</span> <span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="nf">substring</span><span class="p">(</span><span class="nv">from</span><span class="p">:</span> <span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="nf">index</span><span class="p">(</span><span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="n">endIndex</span><span class="p">,</span> <span class="nv">offsetBy</span><span class="p">:</span> <span class="o">-</span><span class="mi">4</span><span class="p">))</span>
</span><span id="L120">    <span class="p">}</span>
</span><span id="L121"><span class="p">}</span>
</span></pre></td></tr></tbody></table>
``` 

i.e. there are 121 lines, and the final closing `}` is wrapped properly. Whereas on the live site as linked here:

```html
<span id="L119">        <span class="k">return</span> <span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="nf">substring</span><span class="p">(</span><span class="nv">from</span><span class="p">:</span> <span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="nf">index</span><span class="p">(</span><span class="k">self</span><span class="o">.</span><span class="n">number</span><span class="o">.</span><span class="n">endIndex</span><span class="p">,</span> <span class="nv">offsetBy</span><span class="p">:</span> <span class="o">-</span><span class="mi">4</span><span class="p">))</span>
</span><span id="L120">    <span class="p">}</span>
</span><span class="p">}</span></pre></td></tr></tbody></table>
```

**Comment by k3rni** [comment]
This may have something to do how the final open newline is closed, by just naively stripping it ouf of the code. But I find it hard to reproduce.

**Comment by kytrinyx** [comment]
If you grab the exact code example and stick it in a test, does it reproduce the issue?

**Comment by k3rni** [comment]
Nope (using the test code I introduced that runs the Python and OCaml samples). The snips above are from exactly that, and ran from 93f9a0.

**Comment by kytrinyx** [comment]
Oh, right. I see what you mean. That's really confusing. How can this be different in production?

**Comment by Insti** [comment]
If this is not going to be fixed quickly, can we revert the change? It makes referring to line numbers in reviews difficult.

**Comment by Insti** [comment]
The tests pass because the test that detect this problem were modified in #3048 (rather than fixing the problem.)

**Comment by Insti** [comment]
I've had a more thorough look through the code and suspect the problem is caused by solutions that do not end with a newline.

I also have some issues with the complexity of the code and overuse of instance variables, particularly in the test code.

I suggest we revert the changes (see #3056) then work on a new PR that addresses the code quality issues and fixes the underlying newline problem. (I am happy to help @k3rni with this if he wants to try again.)

**Comment by Insti** [comment]
> If you grab the exact code example and stick it in a test, does it reproduce the issue?

The answer to this is actually yes.

**Comment by k3rni** [comment]
Could you share that failing test in a gist?

**Comment by k3rni** [comment]
Managed to reproduce it, turns out my editor was lying about the newline following the final `}` not existing (and still saving it). Checking possibilities here.

**Comment by Insti** [comment]
Fixed by #3063 
Thanks for the report @codesman .
